mm2src/common/jsonrpc_client.rs

```diff
@@ -334,133 +350,94 @@ pub trait JsonRpcMultiClient: JsonRpcClient {
         request: JsonRpcRequest,
     ) -> RpcRes<T> {
         let client_info = self.client_info();
+        let request = JsonRpcRequestEnum::Single(request);
         Box::new(
-            self.transport_exact(to_addr.to_owned(), JsonRpcRequestEnum::Single(request.clone()))
-                .then(move |result| process_transport_single_result(result, client_info, request)),
+            self.transport_exact(to_addr.to_owned(), request.clone())
+                .then(move |result| process_transport_single_result(result))
+                .map_err(|e| JsonRpcError::new(client_info, request, e)),
         )
     }
 }
 
 /// Checks if the given `result` is success and contains `JsonRpcResponse`.
 /// Tries to decode the batch response into `T`.
 fn process_transport_single_result<T: DeserializeOwned + Send + 'static>(
-    result: Result<(JsonRpcRemoteAddr, JsonRpcResponseEnum), String>,
-    client_info: String,
-    request: JsonRpcRequest,
-) -> Result<T, JsonRpcError> {
-    let request = JsonRpcRequestEnum::Single(request);
-
+    result: Result<(JsonRpcRemoteAddr, JsonRpcResponseEnum), JsonRpcErrorType>,
+) -> Result<T, JsonRpcErrorType> {
     match result {
-        Ok((remote_addr, JsonRpcResponseEnum::Single(single))) => {
-            process_single_response(client_info, remote_addr, request, single)
-        },
-        Ok((remote_addr, JsonRpcResponseEnum::Batch(batch))) => {
-            let error = ERRL!("Expeced single response, found batch response: {:?}", batch);
-            Err(JsonRpcError {
-                client_info,
-                request,
-                error: JsonRpcErrorType::Parse(remote_addr, error),
-            })
-        },
-        Err(e) => Err(JsonRpcError {
-            client_info,
-            request,
-            error: JsonRpcErrorType::Transport(e),
-        }),
+        Ok((remote_addr, JsonRpcResponseEnum::Single(single))) => process_single_response(remote_addr, single),
+        Ok((remote_addr, JsonRpcResponseEnum::Batch(batch))) => Err(JsonRpcErrorType::Parse(
+            remote_addr,
+            ERRL!("Expected single response, found batch response: {:?}", batch),
```

Comment:
It can take a long time, but what about changing
```rust
pub type JsonRpcResponseFut =
    Box<dyn Future<Item = (JsonRpcRemoteAddr, JsonRpcResponseEnum), Error = JsonRpcErrorType> + Send + 'static>;
```
to
```rust
pub type JsonRpcResponseFut =
    Box<dyn Future<Item = (JsonRpcRemoteAddr, JsonRpcResponseEnum), Error = MmError<JsonRpcErrorType>> + Send + 'static>;
```
Is it even possible? Please stop the researching at the moment if you realize that this refactoring will require a lot of effort.
It will help us to avoid `ERRL` macro

Reply:
@sergeyboyko0791 using MmError in `common` modules causes cyclic deps.. can I make use of generics to make this work ??

```rust
pub type JsonRpcResponseFut =
    Box<dyn Future<Item = (JsonRpcRemoteAddr, JsonRpcResponseEnum), Error =T> + Send + 'static>;
```

Reply:
Then let's postpone this refactoring :)

Reply:
I think it will be overhead to make the `Future::Error` generic. Also it's probably impossible due to `MmError` trait bounds.

Reply:
ok I see